WcaOnRails/lib/paypal_interface.rb

```diff
@@ -0,0 +1,124 @@
+# frozen_string_literal: true
+
+module PaypalInterface
+  def self.generate_paypal_onboarding_link(competition_id)
+    url = "#{EnvConfig.PAYPAL_BASE_URL}/v2/customer/partner-referrals"
+
+    payload = {
+      operations: [
+        {
+          operation: 'API_INTEGRATION',
+          api_integration_preference: {
+            rest_api_integration: {
+              integration_method: 'PAYPAL',
+              integration_type: 'THIRD_PARTY',
+              third_party_details: {
+                features: ['PAYMENT', 'REFUND'],
+              },
+            },
+          },
+        },
+      ],
+      products: ['PPCP'], # TODO: Experiment with other payment types
+      partner_config_override: {
+        return_url: EnvConfig.ROOT_URL + Rails.application.routes.url_helpers.competitions_paypal_return_path(competition_id),
+        return_url_description: "the url to return the WCA after the paypal onboarding process.",
+      },
+      legal_consents: [
+        {
+          type: 'SHARE_DATA_CONSENT',
+          granted: true,
+        },
+      ],
+    }
+
+    response = paypal_connection(url).post do |req|
+      req.body = payload
+    end
+
+    response.body['links'].each do |link|
+      if link['rel'] == "action_url"
+        return link['href']
+      end
+    end
+  end
+
+  def self.create_order(registration)
+    url = "#{EnvConfig.PAYPAL_BASE_URL}/v2/checkout/orders"
+
+    outstanding_fees = registration.outstanding_entry_fees
+    fee_currency = outstanding_fees.currency.iso_code
+
+    payload = {
+      intent: 'CAPTURE',
+      purchase_units: [
+        {
+          amount: { currency_code: fee_currency.to_s, value: outstanding_fees.to_s },
+          # amount: { currency_code: "EUR", value: outstanding_fees.to_s },
+        },
+      ],
+    }
+
+    response = paypal_connection(url).post do |req|
+      req.headers['PayPal-Partner-Attribution-Id'] = AppSecrets.PAYPAL_ATTRIBUTION_CODE
+      req.headers['PayPal-Auth-Assertion'] = get_paypal_auth_assertion(registration.competition)
+
+      req.body = payload
+    end
+
+    response.body
+  end
+
+  def self.capture_payment(competition, order_id)
+    url = "#{EnvConfig.PAYPAL_BASE_URL}/v2/checkout/orders/#{order_id}/capture"
+
+    response = paypal_connection(url).post do |req|
+      req.headers['PayPal-Partner-Attribution-Id'] = AppSecrets.PAYPAL_ATTRIBUTION_CODE
+      req.headers['PayPal-Auth-Assertion'] = get_paypal_auth_assertion(competition)
+    end
+
+    response.body
+  end
+
+  class << self
```

Comment:
Are we happy to use this `class << self` stuff? 

My understanding according to an SO thread that I looked at during the wca-reg repo work is that this is the correct way to define a "private" method that's accessible by a class method (instance methods can't access `private` methods, as far as I understand/have experienced)

Reply:
Hm, not sure about the whole notion of instance methods vs class methods either.
What happens if you dont wrap it into `class << self` and just do `private def self.my:method` instead?

Reply:
Ok prefixing it with `private_class_method` works (just `private` doesn't work) - not sure how I didn't find this in my last research, but good to know! 

(`class << self` is also apparently a valid method, but I think `private_class_method` is WAY more readable/intuitive)